src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberBeginWork
+ * @flow
+ */
+
+'use strict';
+
+import type { ReactCoroutine, ReactYield } from 'ReactCoroutine';
+import type { Fiber } from 'ReactFiber';
+
+var ReactChildFiber = require('ReactChildFiber');
+var ReactTypesOfWork = require('ReactTypesOfWork');
+var {
+  IndeterminateComponent,
+  FunctionalComponent,
+  ClassComponent,
+  HostComponent,
+  CoroutineComponent,
+  YieldComponent,
+} = ReactTypesOfWork;
+
+function getElement(unitOfWork) : ReactElement {
+  var element = unitOfWork.input;
+  if (!element) {
+    throw new Error('Should be resolved by now');
+  }
+  return (element : ReactElement);
+}
+
+function updateFunctionalComponent(unitOfWork) {
+  var element = getElement(unitOfWork);
+  var fn = element.type;
+  var props = element.props;
+  console.log('perform work on:', fn.name);
+  var nextChildren = fn(props);
+
+  unitOfWork.child = ReactChildFiber.reconcileChildFibers(
+    unitOfWork,
+    unitOfWork.child,
+    nextChildren
+  );
+}
+
+function updateHostComponent(unitOfWork) {
+  var element = getElement(unitOfWork);
+  console.log('host component', element.type, typeof element.props.children === 'string' ? element.props.children : '');
+
+  var nextChildren = element.props.children;
+  unitOfWork.child = ReactChildFiber.reconcileChildFibers(
+    unitOfWork,
+    unitOfWork.child,
+    nextChildren
+  );
+}
+
+function mountIndeterminateComponent(unitOfWork) {
+  var element = getElement(unitOfWork);
+  var fn = element.type;
+  var props = element.props;
+  var value = fn(props);
```

Comment:
Oh, I see. `value` is an object only if `fn` is module-pattern-y (like `Relay.createContainer` currently is).